lib/Conversion/ImportVerilog/ImportVerilogInternals.h

```diff
@@ -83,6 +83,10 @@ struct Context {
   /// not been converted yet.
   std::queue<const slang::ast::InstanceBodySymbol *> moduleWorklist;
 
+  /// A list of ports for recording locations of ports to bypass extra
+  /// declaration of net or variable operation.
+  DenseMap<const slang::SourceLocation *, moore::Direction> portMap;
```

Comment:
Using SourceLocation as a key looks slightly counterintuitive to me. Is SourceLocation stable API that we can use in such a way? (I'm not much familiar with slang so please correct me if I'm wrong :) Can we use `const slang::Symbol*` or maybe `StringAttr` instead?

Reply:
Thanks for your review! 🥰 We have used `stringAttr` to construct the container in the dev branch which has the same purpose. It was directly inserted into a scopedHashTable of type `stringAttr` as the primary key. After some adjustments, the type of the original scopedHashtable had been changed due to inappropriate type design, the type of primary key was changed to `slang::ast::ValueSymbol`. It has nothing to do with PortSymbol, MultiPortSymbol, and other symbols. The sourceLocation class represents a location in the source code (or within a macro expression). Every element in AST has a unique location including ports. We even could decode it back to its original locations (line, column, file information in a file location or a macro location)......
The name and location could be both ways to help find when and where we can create the port operation in a Moore module. `slang::ast::Symbol` is the parent of all subdivided symbols defined in slang. Defining the type of the container as `slang::ast::symbol` is sufficient for this purpose. The direction information is still important. It is not accepted because we suspect that the conversions applicable in this particular case should not use the top-level class.

Reply:
@cepheus69 @hailongSun2000  One thing I'm wondering is whether you actually need `moore.port` at the moment for your Moore-to-Core lowering. As far as I understand, Slang always creates nets and variables corresponding to the ports. And the ports feel like they are more about describing the outside-facing interface that you interact with when you instantiate the module, not the nets inside the module. Do you already support module instances in your Moore-to-Core lowering? If not, maybe we can skip `moore.port` for now and figure it out together with the ports/operands on a `moore.instance` operation.

Reply:
> @cepheus69 @hailongSun2000 One thing I'm wondering is whether you actually need `moore.port` at the moment for your Moore-to-Core lowering. As far as I understand, Slang always creates nets and variables corresponding to the ports. And the ports feel like they are more about describing the outside-facing interface that you interact with when you instantiate the module, not the nets inside the module. Do you already support module instances in your Moore-to-Core lowering? If not, maybe we can skip `moore.port` for now and figure it out together with the ports/operands on a `moore.instance` operation.

@cepheus69 expends the original definition of `moore. instance`, we can handle the port connections. IR looks like `moore. instance "bar" @Bar(%a) -> (%b) : (!moore.logic) -> (!moore.logic)`. We also support `moore.module` to `hw.module` and `moore.instance` to `hw.instance`.
The idea is to extend `moore.port` support without changing the code, so we create a new container rather than tweak the `ValueSymbols`.

Reply:
In that case, could we just skip the port AST nodes (not create a `moore.port` op for them), but instead populate the input and output argument list of the `moore.module`, like `moore.module @Bar (%in: type, ...) -> (%out: type, ...)` to match the syntax of the instance?

Later on, as we expand our support for the weird non-ANSI/ANSI ports that Verilog allows, we can lower more complicated port constructs to these simple lists of input and output arguments of the module. We could also think about having only a single list of arguments, and using a special `!moore.out<T>` and `!moore.inout<T>` wrapper type to capture port directions. This would align fairly well with Verilog's model of being able to refer to a port by name and then assigning to it, and instances just being weird ways of translating from instance-site SSA values to module SSA values.

Reply:
If I remember correctly, we can directly get port lists when we handle the module even though skipping the port AST nodes. But we have to expand the definition of the moore::SVModuleOp. Maybe this will take some time to produce a module with a single list of arguments and a special wrapper type to capture the port directions.

Reply:
We could also limit ourselves to simple input and output ports, and make modules use the `(%in, ...) -> (%out, ...)` pattern that you already use for instances. When we see anything that isn't trivially this in the AST, we just error out and inform the user that their port pattern is not supported. We can then expand on this later on.